File: Objective-C/Tests/CBLTestCase.m
//
//  CBLTestCase.m
//  CouchbaseLite
//
//  Copyright (c) 2017 Couchbase, Inc All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "CBLTestCase.h"
#include "c4.h"
#import "CollectionUtils.h"
#import "CBLLogSinks+Reset.h"

#ifdef COUCHBASE_ENTERPRISE
#define kDatabaseDirName @"CouchbaseLite_EE"
#else
#define kDatabaseDirName @"CouchbaseLite"
#endif

const NSTimeInterval kExpTimeout = 20.0;

@implementation CBLTestCase
{
    int _c4ObjectCount;
}

@synthesize db=_db, otherDB=_otherDB, disableObjectLeakCheck=_disableObjectLeakCheck;

// TODO: Remove https://issues.couchbase.com/browse/CBL-3206
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"

- (XCTestExpectation*) allowOverfillExpectationWithDescription:(NSString *)description {
    XCTestExpectation* e = [super expectationWithDescription: description];
    e.assertForOverFulfill = false;
    return e;
}

- (void) setUp {
    [super setUp];
    
    [self deleteDBNamed: kDatabaseName error: nil];
    [self deleteDBNamed: kOtherDatabaseName error: nil];
    
    _c4ObjectCount = c4_getObjectCount();
    NSString* dir = self.directory;
    if ([[NSFileManager defaultManager] fileExistsAtPath: dir]) {
        NSError* error;
        Assert([[NSFileManager defaultManager] removeItemAtPath: dir error: &error],
               @"Error deleting CouchbaseLite folder: %@", error);
    }
    [self initDB];
    
    // Debug logging to be made before this, if any
    [CBLLogSinks resetApiVersion];
}

- (void) tearDown {
    [CBLLogSinks resetApiVersion];
    
    if (_db) {
        NSError* error;
        Assert([_db close: &error], @"Failed to close db: %@", error);
        _db = nil;
    }
    
    if (_otherDB) {
        NSError* error;
        Assert([_otherDB close: &error], @"Failed to close otherdb: %@", error);
        _otherDB = nil;
    }
    
    if (!_disableObjectLeakCheck) {
        // Wait a little while for objects to be cleaned up:
        int leaks = 0;
        for (int i = 0; i < 20; i++) {
            leaks = c4_getObjectCount() - _c4ObjectCount;
            if (leaks == 0)
                break;
            else
                [NSThread sleepForTimeInterval: 0.1];
        }
        if (leaks) {
            fprintf(stderr, "**** LITECORE OBJECTS STILL NOT FREED: ****\n");
            c4_dumpInstances();
            XCTFail("%d LiteCore objects have not been freed (see above)", leaks);
        }
    }
    
    [super tearDown];
}

- (NSString*) directory {
    return [NSTemporaryDirectory() stringByAppendingPathComponent: kDatabaseDirName];
}

- (BOOL) hasHostApp {
#if TARGET_OS_IPHONE
    NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
    return [defaults boolForKey: @"hostApp"];
#else
    return YES;
#endif
}

- (BOOL) keyChainAccessAllowed {
    #if TARGET_OS_IPHONE
        return self.hasHostApp;
    #else
        return YES;
    #endif
}

- (NSString*) databasePath: (NSString*)fileName inDirectory: (NSString*)dir {
    NSString *directory = [@"Support/databases" stringByAppendingPathComponent:dir];
    NSString* path = [[NSBundle bundleForClass: [self class]] pathForResource: fileName
                                                                       ofType: nil
                                                                  inDirectory: directory];
    Assert(path, @"FATAL: Missing file '%@' in bundle directory '%@'", fileName, directory);
    return path;
}

- (CBLDatabase*) openDBNamed: (NSString*)name error: (NSError**)error {
    CBLDatabaseConfiguration* config = [[CBLDatabaseConfiguration alloc] init];
    config.directory = self.directory;
    return [[CBLDatabase alloc] initWithName: name config: config error: error];
}

- (void) initDB {
    [self openDB];
}

- (void) openDB {
    Assert(!_db);
    NSError* error;
    _db = [self openDBNamed: kDatabaseName error: &error];
    AssertNil(error);
    AssertNotNil(_db);
}

- (void) reopenDB {
    NSError *error;
    if (_db) {
        Assert([_db close: &error], @"Close error: %@", error);
        _db = nil;
    }
    [self openDB];
}

- (void) cleanDB {
    NSError *error;
    Assert([_db delete: &error], @"Delete error: %@", error);
    [self reopenDB];
}

- (void) openOtherDB {
    Assert(!_otherDB);
    NSError* error;
    _otherDB = [self openDBNamed: kOtherDatabaseName error: &error];
    AssertNil(error);
    AssertNotNil(_db);
}

- (void) reopenOtherDB {
    if (_otherDB) {
        NSError *error;
        Assert([_otherDB close: &error], @"Close error: %@", error);
        AssertNil(error);
        _otherDB = nil;
    }
    [self openOtherDB];
}

- (BOOL) deleteDBNamed: (NSString*)name error: (NSError**)error {
    __block BOOL result;
    __block NSError* err = nil;
    [self ignoreExceptionBreakPoint: ^{
        result = [CBLDatabase deleteDatabase: name inDirectory: self.directory error: &err];
    }];
    if (error) {
        *error = err;
    }
    return result;
}

- (void) deleteDatabase: (CBLDatabase*)database {
    NSError* error;
    NSString* path = database.path;
    Assert([[NSFileManager defaultManager] fileExistsAtPath: path]);
    Assert([database delete: &error]);
    AssertNil(error);
    AssertFalse([[NSFileManager defaultManager] fileExistsAtPath: path]);
}

- (void) closeDatabase: (CBLDatabase*)database{ 
    NSError* error;
    Assert([database close:&error]);
    AssertNil(error);
}

- (CBLMutableDocument*) createDocument {
    return [[CBLMutableDocument alloc] init];
}

- (CBLMutableDocument*) createDocument: (NSString*)documentID {
    return [[CBLMutableDocument alloc] initWithID: documentID];
}

- (CBLMutableDocument*) createDocument:(NSString *)documentID data:(NSDictionary *)data {
    return [[CBLMutableDocument alloc] initWithID: documentID data: data];
}

- (CBLMutableDocument*) generateDocumentWithID: (NSString*)documentID {
    CBLMutableDocument* doc = [self createDocument: documentID];
    [doc setValue: @1 forKey:@"key"];
    [self saveDocument: doc];
    AssertEqual(doc.sequence, 1u);
    if (documentID)
        AssertEqualObjects(doc.id, documentID);
    return doc;
}

- (void) saveDocument: (CBLMutableDocument*)document {
    [self saveDocument: document collection: nil];
}

- (void) saveDocument:(CBLMutableDocument *)document collection: (nullable CBLCollection*)col {
    NSError* error;
    CBLCollection* c = col ?: [self.db defaultCollection: &error];
    AssertNotNil(c);
    
    Assert([c saveDocument: document error: &error], @"Saving error: %@", error);
    AssertNil(error);
    
    CBLDocument* savedDoc = [c documentWithID: document.id error: &error];
    AssertNotNil(savedDoc);
    AssertEqualObjects(savedDoc.id, document.id);
    AssertEqualObjects([savedDoc toDictionary], [document toDictionary]);
    AssertNil(error);
}

- (void) saveDocument: (CBLMutableDocument*)document eval: (void(^)(CBLDocument*))block {
    NSError* error;
    block(document);
    Assert([_db saveDocument: document error: &error], @"Saving error: %@", error);
    block(document);
    block([_db documentWithID: document.id]);
}

- (void) createDocNumbered: (nullable CBLCollection*)col start: (NSInteger)start num: (NSInteger)num {
    for (NSInteger i = start; i < (start+num); i++) {
        NSString* docID = [NSString stringWithFormat: @"doc%ld", (long)i];
        CBLMutableDocument* doc = [[CBLMutableDocument alloc] initWithID: docID];
        [doc setValue: @(i) forKey: @"number1"];
        [doc setValue: @(num-i) forKey: @"number2"];
        [self saveDocument: doc collection: col];
    }
}

- (NSURL*) urlForResource: (NSString*)resourceName ofType: (NSString*)type {
    NSString* res = [@"Support" stringByAppendingPathComponent: resourceName];
    return [[NSBundle bundleForClass: [self class]] URLForResource: res withExtension: type];
}

- (NSData*) dataFromResource: (NSString*)resourceName ofType: (NSString*)type {
    NSString* res = [@"Support" stringByAppendingPathComponent: resourceName];
    NSString* path = [[NSBundle bundleForClass: [self class]] pathForResource: res
                                                                       ofType: type];
    Assert(path, @"Missing test file %@.%@", resourceName, type);
    NSData* contents = [NSData dataWithContentsOfFile: path
                                              options: 0
                                                error: NULL];
    Assert(contents);
    return contents;
}

- (NSString*) stringFromResource: (NSString*)resourceName ofType: (NSString*)type {
    NSData* contents = [self dataFromResource: resourceName ofType: type];
    NSString* str = [[NSString alloc] initWithData: contents
                                          encoding: NSUTF8StringEncoding];
    Assert(str);
    return str;
}

- (NSString*) randomStringWithLength: (NSUInteger)length {
    static NSString *chars = @"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXZY0123456789";
    NSMutableString *s = [NSMutableString stringWithCapacity: length];
    for (NSUInteger i = 0; i < length; i++) {
        [s appendFormat:@"%C",
            [chars characterAtIndex: (arc4random() % [chars length])]];
    }
    return s;
}

- (void) loadJSONResource: (NSString*)resourceName {
    NSError* error = nil;
    CBLCollection* defaultCollection = [self.db defaultCollection: &error];
    AssertNil(error);
    
    [self loadJSONResource: resourceName toCollection: defaultCollection];
    
}

- (void) loadJSONResource: (NSString*)resourceName toCollection: (CBLCollection*)collection {
    @autoreleasepool {
        NSString* contents = [self stringFromResource: resourceName ofType: @"json"];
        return [self loadJSONString: contents named: resourceName toCollection: collection];
    }
}

- (void) loadJSONString: (NSString*)contents named: (NSString*)resourceName {
    NSError* error = nil;
    CBLCollection* defaultCollection = [self.db defaultCollection: &error];
    AssertNil(error);
    
    [self loadJSONString: contents named: resourceName toCollection: defaultCollection];
}

- (void) loadJSONString: (NSString*)contents
                  named: (NSString*)resourceName
           toCollection: (CBLCollection*)collection {
    @autoreleasepool {
        __block uint64_t n = 0;
        [contents enumerateLinesUsingBlock: ^(NSString *line, BOOL *stop) {
            NSError* err;
            CBLMutableDocument* doc = [[CBLMutableDocument alloc] initWithID: $sprintf(@"doc-%03llu", ++n)
                                                                        json: line error: &err];
            Assert([collection saveDocument: doc error: &err], @"Couldn't save document: %@", err);
        }];
    }
}

- (CBLBlob*) blobForString: (NSString*)string {
    return [[CBLBlob alloc] initWithContentType: @"text/plain"
                                           data: [string dataUsingEncoding: NSUTF8StringEncoding]];
}

- (SecCertificateRef) createSecCertFromPEM: (NSString*)pem {
    // Separate the lines and remove BEGIN/END headers
    NSMutableArray<NSString*>*lines = [NSMutableArray array];
    for (NSString* line in [pem componentsSeparatedByCharactersInSet: [NSCharacterSet newlineCharacterSet]]) {
        if (![line containsString: @"BEGIN CERTIFICATE"] && ![line containsString: @"END CERTIFICATE"]) {
            [lines addObject:line];
        }
    }
    
    // Join the base64 lines
    NSString* base64String = [lines componentsJoinedByString: @""];
    
    // Decode base64 to NSData
    NSData* certData = [[NSData alloc] initWithBase64EncodedString: base64String options: 0];
    AssertNotNil(certData);
    
    return SecCertificateCreateWithData(NULL, (__bridge CFDataRef)certData);
}

// helper method to check error
- (void) expectError: (NSErrorDomain)domain code: (NSInteger)code in: (BOOL (^)(NSError**))block {
    if ([self isProfiling])
        return;
    
    ++gC4ExpectExceptions;
    NSError* error;
    BOOL succeeded = block(&error);
    --gC4ExpectExceptions;

    if (succeeded) {
        XCTFail("Block expected to fail but didn't");
    } else {
        XCTAssert([domain isEqualToString: error.domain] && code == error.code,
                  "Block expected to return error (%@ %ld), but instead returned %@",
                  domain, (long)code, error);
    }
}

- (void) expectException: (NSString*)name in: (void (^) (void))block {
    if ([self isProfiling])
        return;
    
    ++gC4ExpectExceptions;
    XCTAssertThrowsSpecificNamed(block(), NSException, name);
    --gC4ExpectExceptions;
}

- (void) mayHaveException: (NSString*)name in: (void (^) (void))block {
    if ([self isProfiling])
        return;
    
    @try {
        ++gC4ExpectExceptions;
        block();
    }
    @catch (NSException* e) {
        AssertEqualObjects(e.name, name);
    }
    @finally {
        --gC4ExpectExceptions;
    }
}

- (void) ignoreException: (void (^) (void))block {
    if ([self isProfiling])
        return;
    
    @try {
        ++gC4ExpectExceptions;
        block();
    }
    @catch (NSException* e) { }
    @finally {
        --gC4ExpectExceptions;
    }
}

- (void) ignoreExceptionBreakPoint: (void (^) (void))block {
    ++gC4ExpectExceptions;
    block();
    --gC4ExpectExceptions;
}

- (uint64_t) verifyQuery: (CBLQuery*)q
            randomAccess: (BOOL)randomAccess
                    test: (void (^)(uint64_t n, CBLQueryResult *result))block {
    NSError* error;
    CBLQueryResultSet* rs = [q execute: &error];
    Assert(rs, @"Query failed: %@", error);
    uint64_t n = 0;
    for (CBLQueryResult *r in rs) {
        block(++n, r);
    }
    
    rs = [q execute: &error];
    Assert(rs, @"Query failed: %@", error);
    NSArray* all = rs.allObjects;
    AssertEqual(all.count, n);
    if (randomAccess && n > 0) {
        // Note: the block's 1st parameter is 1-based, while NSArray is 0-based
        block(n,       all[(NSUInteger)(n-1)]);
        block(1,       all[0]);
        block(n/2 + 1, all[(NSUInteger)(n/2)]);
    }
    return n;
}

- (BOOL) isUsingIndexNamed: (NSString*)indexName forQuery: (CBLQuery*)query {
    NSError* error;
    NSString* plan = [query explain: &error];
    AssertNil(error);
    AssertNotNil(plan);
    
    NSString *usingIndex = [NSString stringWithFormat:@"USING INDEX %@", indexName];
    NSString *usingCoveringIndex = [NSString stringWithFormat:@"USING COVERING INDEX %@", indexName];
    
    return ([plan rangeOfString:usingIndex].location != NSNotFound ||
            [plan rangeOfString:usingCoveringIndex].location != NSNotFound);
}

- (BOOL) isProfiling {
    return NSProcessInfo.processInfo.environment[@"PROFILING"] != nil;
}

- (NSString*) getRickAndMortyJSON {
    NSError* err;
    NSData* content = [@"Earth(C-137)" dataUsingEncoding: NSUTF8StringEncoding];
    CBLBlob* blob = [[CBLBlob alloc] initWithContentType:@"text/plain" data: content];
    [self.db saveBlob: blob error: &err];
    
    content = [@"Grandpa Rick" dataUsingEncoding: NSUTF8StringEncoding];
    blob = [[CBLBlob alloc] initWithContentType:@"text/plain" data: content];
    [self.db saveBlob: blob error: &err];
    
    return [self stringFromResource: @"rick_morty" ofType: @"json"];
}

#pragma clang diagnostic pop

@end

File: Swift/Tests/CBLTestCase.swift
//
//  CBLTestCase.swift
//  CouchbaseLite
//
//  Copyright (c) 2017 Couchbase, Inc All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import XCTest
import Foundation
@testable import CouchbaseLiteSwift

extension String {
    func toJSONObj() -> Any {
        let d = self.data(using: .utf8)!
        return try! JSONSerialization.jsonObject(with: d, options: [])
    }
}

class CBLTestCase: XCTestCase {
    let expTimeout: TimeInterval = 20.0

    /// Opened when setting up each test case.
    var db: Database!
    
    /// Need to explicitly open by calling openOtherDB() function.
    var otherDB: Database?
    
    let databaseName = "testdb"
    
    let otherDatabaseName = "otherdb"
    
    var defaultCollection: Collection?
    
    var otherDB_defaultCollection: Collection?
    
    #if COUCHBASE_ENTERPRISE
        let directory = NSTemporaryDirectory().appending("CouchbaseLite-EE")
    #else
        let directory = NSTemporaryDirectory().appending("CouchbaseLite")
    #endif
    
    var isHostApp: Bool {
    #if os(iOS)
        let defaults = UserDefaults.standard
        return defaults.bool(forKey: "hostApp")
    #else
        return true
    #endif
    }
    
    var keyChainAccessAllowed: Bool {
    #if os(iOS)
        return self.isHostApp
    #else
        return true
    #endif
    }
    
    /// This expectation will allow overfill expectation.
    /// CBL-2363: Replicator might send extra idle status when its being stopped, which is not a bug
    func allowOverfillExpectation(description: String) -> XCTestExpectation {
        let e = super.expectation(description: description)
        e.assertForOverFulfill = false
        return e
    }
    
    override func setUp() {
        super.setUp()
        
        try? deleteDB(name: databaseName);
        
        try? deleteDB(name: otherDatabaseName);
        
        if FileManager.default.fileExists(atPath: self.directory) {
            try! FileManager.default.removeItem(atPath: self.directory)
        }
        XCTAssertTrue(!FileManager.default.fileExists(atPath: self.directory))
        
        try! initDB()
        
        LogSinks._resetApiVersion()
    }
    
    override func tearDown() {
        self.defaultCollection = nil
        self.otherDB_defaultCollection = nil
        try! db.close()
        try! otherDB?.close()
        
        LogSinks._resetApiVersion()
        
        super.tearDown()
    }
    
    func openDB(name: String) throws -> Database {
        var config = DatabaseConfiguration()
        config.directory = self.directory
        return try Database(name: name, config: config)
    }
    
    func openDB() throws {
        db = try openDB(name: databaseName)
        self.defaultCollection = try! db.defaultCollection()
    }
    
    func initDB() throws {
        try openDB()
    }
    
    func reopenDB() throws {
        try db.close()
        db = nil
        try openDB()
    }
    
    func cleanDB() throws {
        try db.delete()
        try reopenDB()
    }
    
    func openOtherDB() throws {
        otherDB = try openDB(name: otherDatabaseName)
        self.otherDB_defaultCollection = try! otherDB?.defaultCollection()
    }
    
    func reopenOtherDB() throws {
        try otherDB?.close()
        otherDB = nil
        try openOtherDB()
    }
    
    func deleteDB(name: String) throws {
        ignoreException {
            try Database.delete(withName: name, inDirectory: self.directory)
        }
    }
    
    func createDocument() -> MutableDocument {
        return MutableDocument()
    }
    
    func createDocument(_ id: String?) -> MutableDocument {
        return MutableDocument(id: id)
    }
    
    func createDocument(data: [String:Any]) -> MutableDocument {
        return MutableDocument(data: data)
    }
    
    func createDocument(_ id: String?, data: [String:Any]) -> MutableDocument {
        return MutableDocument(id: id, data: data)
    }
    
    func createDocNumbered(_ col: Collection, start: Int, num: Int) throws {
        for i in start..<start+num {
            let mdoc = createDocument("doc\(i)")
            mdoc.setInt(i, forKey: "number1")
            try col.save(document: mdoc)
        }
    }
    
    @discardableResult
    func generateDocument(withID id: String?) throws -> MutableDocument {
        let doc = createDocument(id);
        doc.setValue(1, forKey: "key")
        try saveDocument(doc)
        XCTAssertEqual(doc.sequence, 1)
        XCTAssertNotNil(doc.id)
        if id != nil {
            XCTAssertEqual(doc.id, id)
        }
        return doc
    }
    
    func saveDocument(_ document: MutableDocument) throws {
        try defaultCollection!.save(document: document)
        let savedDoc = try defaultCollection!.document(id: document.id)
        XCTAssertNotNil(savedDoc)
        XCTAssertEqual(savedDoc!.id, document.id)
    }
    
    func saveDocument(_ document: MutableDocument, eval: (Document) -> Void) throws {
        eval(document)
        try saveDocument(document)
        eval(document)
        let savedDoc = try defaultCollection!.document(id: document.id)!
        eval(savedDoc)
    }
    
    func saveDocument(_ document: MutableDocument, collection: Collection) throws {
        try collection.save(document: document)
        let savedDoc = try collection.document(id: document.id)
        XCTAssertNotNil(savedDoc)
        XCTAssertEqual(savedDoc!.id, document.id)
    }
    
    func urlForResource(name: String, ofType type: String) -> URL? {
        let res = ("Support" as NSString).appendingPathComponent(name)
        return Bundle(for: Swift.type(of:self)).url(forResource: res, withExtension: type)
    }
    
    func dataFromResource(name: String, ofType type: String) throws -> Data {
        let res = ("Support" as NSString).appendingPathComponent(name)
        let path = Bundle(for: Swift.type(of:self)).path(forResource: res, ofType: type)
        return try! NSData(contentsOfFile: path!, options: []) as Data
    }

    func stringFromResource(name: String, ofType type: String) throws -> String {
        let res = ("Support" as NSString).appendingPathComponent(name)
        let path = Bundle(for: Swift.type(of:self)).path(forResource: res, ofType: type)
        return try String(contentsOfFile: path!, encoding: String.Encoding.utf8)
    }
    
    func decodeFromJSONResource<T: Decodable>(_ name: String, as type: T.Type, limit: Int? = nil) throws -> Array<T> {
        let contents = try stringFromResource(name: name, ofType: "json")
        var n = 0
        var result: Array<T> = []
        contents.enumerateLines(invoking: { (line: String, stop: inout Bool) in
            n += 1
            let decoder = JSONDecoder()
            result.append(try! decoder.decode(T.self, from: line.data(using: String.Encoding.utf8)!))
            if n == limit {
                stop = true
            }
        })
        return result
    }
    
    func loadJSONResource(_ name: String, collection: Collection) throws {
        try loadJSONResource(name, collection: collection, limit: Int.max)
    }
    
    func loadJSONResource(_ name: String, collection: Collection, limit: Int, idKey: String? = nil) throws {
        try autoreleasepool {
            let contents = try stringFromResource(name: name, ofType: "json")
            var n = 0
            contents.enumerateLines(invoking: { (line: String, stop: inout Bool) in
                n += 1
                let json = line.data(using: String.Encoding.utf8, allowLossyConversion: false)
                var dict = try! JSONSerialization.jsonObject(with: json!, options: []) as! [String:Any]
                let docID: String
                if let idKey = idKey {
                    docID = dict.removeValue(forKey: idKey) as! String
                } else {
                    docID = String(format: "doc-%03llu", n)
                }
                let doc = MutableDocument(id: docID, data: dict)
                try! collection.save(document: doc)
                if n == limit {
                    stop = true
                }
            })
        }
    }
    
    func loadJSONResource(name: String) throws {
        try loadJSONResource(name, collection: self.defaultCollection!)
    }
    
    func jsonFromDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
        formatter.timeZone = NSTimeZone(abbreviation: "UTC")! as TimeZone
        return formatter.string(from: date).appending("Z")
    }
    
    func dateFromJson(_ date: String) -> Date {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZ"
        formatter.timeZone = NSTimeZone.local
        return formatter.date(from: date)!
    }
    
    func blobForString(_ string: String) -> Blob {
        let data = string.data(using: .utf8)!
        return Blob(contentType: "text/plain", data: data)
    }
    
    func expectError(domain: String, code: Int, block: @escaping () throws -> Void) {
        CBLTestHelper.allowException {
            var error: NSError?
            do {
                try block()
            }
            catch let e as NSError {
                error = e
            }
            
            XCTAssertNotNil(error, "Block expected to fail but didn't")
            XCTAssertEqual(error?.domain, domain)
            XCTAssertEqual(error?.code, code)
        }
    }
    
    func expectException(exception: NSExceptionName, block: @escaping () -> Void) {
        var exceptionThrown = false
        do {
            try CBLTestHelper.catchException {
                block()
            }
        } catch {
            XCTAssertEqual((error as NSError).domain, exception.rawValue)
            exceptionThrown = true
        }
        
        XCTAssert(exceptionThrown, "No exception thrown")
    }
    
    func ignoreException(block: @escaping () throws -> Void) {
        CBLTestHelper.allowException {
            try? block()
        }
    }
    
    @discardableResult
    func verifyQuery(_ query: Query, block: (UInt64, Result) throws ->Void) throws -> UInt64 {
        var n: UInt64 = 0
        for row in try query.execute() {
            n += 1
            try block(n, row)
        }
        return n
    }
    
    func isUsingIndex(named indexName: String, for query: Query) throws -> Bool {
        let plan = try query.explain()
        let usingIndex = "USING INDEX \(indexName)"
        let usingCoveringIndex = "USING COVERING INDEX \(indexName)"
        return plan.contains(usingIndex) || plan.contains(usingCoveringIndex)
    }
    
    func getRickAndMortyJSON() throws -> String {
        var content = "Earth(C-137)".data(using: .utf8)!
        var blob = Blob(contentType: "text/plain", data: content)
        try self.db.saveBlob(blob: blob)
        
        content = "Grandpa Rick".data(using: .utf8)!
        blob = Blob(contentType: "text/plain", data: content)
        try self.db.saveBlob(blob: blob)
        
        return try stringFromResource(name: "rick_morty", ofType: "json")
    }
    
    func createSecCertFromPEM(_ pem: String) throws -> SecCertificate {
        // Split lines and filter out header/footer
        let lines = pem
            .components(separatedBy: .newlines)
            .filter { !$0.contains("BEGIN CERTIFICATE") && !$0.contains("END CERTIFICATE") }
        
        // Join base64 lines
        let base64String = lines.joined()
        
        // Decode base64 to Data
        guard let certData = Data(base64Encoded: base64String) else {
            throw NSError(domain: "CertDecode", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 certificate"])
        }
        
        // Create SecCertificate
        guard let cert = SecCertificateCreateWithData(nil, certData as CFData) else {
            throw NSError(domain: "CertDecode", code: -2, userInfo: [NSLocalizedDescriptionKey: "Failed to create SecCertificate"])
        }
        
        return cert
    }
    
}

/** Comparing JSON Dictionary */
public func ==(lhs: [String: Any], rhs: [String: Any] ) -> Bool {
    return NSDictionary(dictionary: lhs).isEqual(to: rhs)
}

/** Comparing JSON Array */
public func ==(lhs: [Any], rhs: [Any] ) -> Bool {
    return NSArray(array: lhs).isEqual(to: rhs)
}
